codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt

```diff
@@ -0,0 +1,206 @@
+    val resourceOperationShapes = service
+        .resources
+        .mapNotNull { model.getShape(it).orNull() }
+        .mapNotNull { it as? ResourceShape }
+        .flatMap { it.allOperations }
+        .mapNotNull { model.getShape(it).orNull() }
+        .mapNotNull { it as? OperationShape }
+    val operationShapes = service.operations.mapNotNull { model.getShape(it).orNull() }.mapNotNull { it as? OperationShape }
+    return resourceOperationShapes + operationShapes
+}
+
+interface ServerProtocol {
+    // Returns the core `Protocol`.
+    fun coreProtocol(): Protocol
+
+    // Returns the Rust marker struct enjoying `OperationShape`.
+    fun markerStruct(): RuntimeType
+
+    // Returns the Rust router type.
+    fun routerType(): RuntimeType
+
+    // Returns the construction of the `routerType` given a `ServiceShape`, a collection of operation values
+    // (`self.operation_name`, ...), and the `Model`.
+    // TODO(Decouple): Perhaps this should lean on a Rust interface.
+    fun routerConstruction(service: ServiceShape, operationValues: Iterable<Writable>, model: Model): Writable
+
+    companion object {
+        // Upgrades the core protocol to a `ServerProtocol`.
+        fun fromCoreProtocol(coreCodegenContext: CoreCodegenContext, protocol: Protocol): ServerProtocol {
+            val serverProtocol = when (protocol) {
+                is AwsJson -> ServerAwsJsonProtocol(coreCodegenContext, protocol)
+                is RestJson -> ServerRestJsonProtocol(coreCodegenContext, protocol)
+                is RestXml -> ServerRestXmlProtocol(coreCodegenContext, protocol)
+                else -> throw IllegalStateException("unsupported protocol")
+            }
+            return serverProtocol
+        }
+    }
+}
+
+class ServerAwsJsonProtocol(
+    coreCodegenContext: CoreCodegenContext,
+    private val coreProtocol: AwsJson,
+) : ServerProtocol {
+    private val runtimeConfig = coreCodegenContext.runtimeConfig
+    private val codegenScope = arrayOf(
+        "SmithyHttpServer" to ServerCargoDependency.SmithyHttpServer(runtimeConfig).asType(),
+    )
+    private val symbolProvider = coreCodegenContext.symbolProvider
+
+    override fun coreProtocol(): Protocol {
+        return coreProtocol
+    }
+
+    override fun markerStruct(): RuntimeType {
+        val name = when (coreProtocol.version) {
+            is AwsJsonVersion.Json10 -> {
+                "AwsJson10"
+            }
+            is AwsJsonVersion.Json11 -> {
+                "AwsJson11"
+            }
+            else -> throw IllegalStateException()
+        }
+        return RuntimeType(name, ServerCargoDependency.SmithyHttpServer(runtimeConfig), "${runtimeConfig.crateSrcPrefix}_http_server::protocols")
+    }
+
+    override fun routerType(): RuntimeType {
+        return RuntimeType("AwsJsonRouter", ServerCargoDependency.SmithyHttpServer(runtimeConfig), "${runtimeConfig.crateSrcPrefix}_http_server::routing::routers::aws_json")
+    }
```

Comment:
```suggestion
    override fun routerType() = RuntimeType("AwsJsonRouter", ...
```

Reply:
You can define one-line functions without creating a `{}` block like this. There are several places in the diff, especially in this file, where you could use this to reduce verbosity a bit.